packages/optimistic-oracle/src/proposer.js

```diff
@@ -73,60 +73,209 @@ class OptimisticOracleProposer {
+  async sendDisputes() {
+    this.logger.debug({
+      at: "OptimisticOracleProposer#sendDisputes",
+      message: "Checking for undisputed price requests to dispute"
+    });
 
-      // With pricefeed successfully constructed, get a proposal price
-      await priceFeed.update();
-      let proposalPrice;
-      try {
-        proposalPrice = (await priceFeed.getHistoricalPrice(priceRequest.timestamp)).toString();
-      } catch (error) {
-        this.logger.error({
-          at: "OptimisticOracleProposer#sendProposals",
-          message: "Failed to query historical price for price request",
-          priceRequest,
-          error
-        });
-        continue;
-      }
+    for (let priceRequest of this.optimisticOracleClient.getUndisputedProposals()) {
+      await this._sendDispute(priceRequest);
+    }
+  }
+
+  // Settle disputes where this bot was the disputer and proposals where this bot was the proposer.
+  async settleRequests() {
+    this.logger.debug({
+      at: "OptimisticOracleProposer#settleRequests",
+      message: "Checking for proposals and disputes to settle"
+    });
+
+    for (let priceRequest of this.optimisticOracleClient
+      .getSettleableProposals(this.account)
+      .concat(this.optimisticOracleClient.getSettleableDisputes(this.account))) {
+      await this._settleRequest(priceRequest);
+    }
+  }
+
+  /** **********************************
+   *
+   * INTERNAL METHODS
+   *
+   ************************************/
+
+  // Construct proposal transaction and send or return early if an error is encountered.
+  async _sendProposal(priceRequest) {
+    const priceFeed = await this._createOrGetCachedPriceFeed(priceRequest.identifier);
+
+    // Pricefeed is either constructed correctly or is null.
+    if (!priceFeed) {
+      this.logger.error({
+        at: "OptimisticOracleProposer#sendProposals",
+        message: "Failed to construct a PriceFeed for price request",
+        priceRequest
+      });
+      return;
+    }
+
+    // With pricefeed successfully constructed, get a proposal price
+    await priceFeed.update();
+    let proposalPrice;
+    try {
+      proposalPrice = (await priceFeed.getHistoricalPrice(priceRequest.timestamp)).toString();
+    } catch (error) {
+      this.logger.error({
+        at: "OptimisticOracleProposer#sendProposals",
+        message: "Failed to query historical price for price request",
+        priceRequest,
+        error
+      });
+      return;
+    }
+
+    // Create the transaction.
+    const proposal = this.ooContract.methods.proposePrice(
+      priceRequest.requester,
+      this.utf8ToHex(priceRequest.identifier),
+      priceRequest.timestamp,
+      priceRequest.ancillaryData,
+      proposalPrice
+    );
+
+    // Simple version of inventory management: simulate the transaction and assume that if it fails,
+    // the caller didn't have enough collateral.
+    let proposalBond, gasEstimation;
+    try {
```

Comment:
Not something we should do in this PR, but some of this transaction sending logic seems like it could be generalized and tossed into a library that could be used here and in the other bots since we do this in so many places. There are _some_ differences in the logic that make this a little difficult, but I could basically imagine a function that's called like this:
```js
const proposal = this.ooContract.methods.proposePrice(
      priceRequest.requester,
      this.utf8ToHex(priceRequest.identifier),
      priceRequest.timestamp,
      priceRequest.ancillaryData,
      proposalPrice
    );
await runTransaction(proposal).catch(e => {
  if (e.type === "call") {
    logger.error({
      ...
    });
  } else if (e.type === "send") {
    logger.error({
      ...
    });
  }
});
```

Reply:
Will do in next PR.